### PR TITLE
Implement isGradingPublished (plus related features)

### LINKED ISSUE #746: Grading status after submission deadline, before grading
Currently, we have:

 'Attempted' means the student has saved for all the sub-questions, 
'Attempting' means the student has saved for some but not all of the questions, and 
'Submitted' means the student actually clicked the submit button.

In these three cases, the grading status should be "not graded". Currently, "Attempted" and "Attempting" gives grading status "not applicable" when mouseover.

### LINKED ISSUE #799: Grading: "Save" allows students to access unfinished grades
Sometimes, grading can take a while or involve multiple graders, and graders will save their work to the database to resume grading at a later time. However, this actually exposes the unfinished work to the students as well.

They may get confused when their grades change constantly, so making grades available for students to view should feel more permanent, especially with the notification system. Once the grades are made available for view by the students, changes should only be made if there is a good reason to, like to fix mistakes or to carry out changes in the marking scheme

Refer to the discussion on pull request #789 and issue #739

### LINKED ISSUE #2189: Visibility of grading icons for resubmissions
Currently, the grading icons are green ticks when the students resubmit a previously graded mission or quest.
This makes it hard to differentiate resubmissions that require attention to the graders, from the graded submissions.

It will be nice if resubmissions that are yet to be re-graded can have a different grading icon from the graded submissions. For instance, the grading icon should be changed to a yellow tick when the avenger unsubmits the mission. They should be changed to green color only after the resubmission is graded.

### PR DESCRIPTION
**Pull request by josh1248**
### Description
Paired with backend request https://github.com/source-academy/backend/pull/965

Closes #746
Closes #799
Closes #2189
Mitigates #2522 (reduces but does not eliminate issue of graded, resubmitted work outright)
Affects #738 (grading status has been revamped, and also no longer considers partially graded work as a separate category)
Supersedes #2546

## Summary of changes

- Introduces isGradingPublished as specified on #2522 into the gradings page, but with modifications to support autograded submissions (solves #799, solves #2189, mitigates #2522).
- Introduces publishAll and unpublishAll in GroundControl (admins only), allowing for batch publishing.
- Introduces isAutoPublished that permits autograded submissions to be automatically published by the autograder
- Revamp columns in grading submissions table: combine 2 columns + publishing status into a single column for simplified tracking (solves #746, affects #738).
- Removes avenger need to track autograded work
- Tweak notifications to reflect published and unpublished actions instead of graded / ungraded actions

image

^(Note: this is a display on tanstack, and will look a bit different after the imminent migration to ag-grid under https://github.com/source-academy/frontend/pull/2893.) slide shows the new 1-column category, "progress", with its varying statuses and allowed actions within the grading submissions table. publishing for graded submissions are activated. Avengers cannot affect autograded work, except for emergency XP changes if they need to.

image

^The new buttons under 'Release grading' allow admins to bulk-publish or bulk-unpublish any assessment. This is a quality-of-life feature for manually graded assessments, but a necessary feature for autograded assessments.

### Introduce isGradingPublished
This pull request implements the isGradingPublished feature as described on #2522, but in a modified manner. The original intention was focused on manually graded work which had been resubmitted. This pull request introduces this new field to the frontend to help in this - grading will not be shown until publishing. However, this pull request additionally implements the published category for autograded assessments. This is to allow for cases when hiding autograder results is be desirable, such as for PEs.

image

^New interface for grading submissions showing implementation of isGradingPublished. (Note: this is a display on tanstack, and will look a bit different after the imminent migration to ag-grid under #2893.)

image

^Even though mission has been fully graded, student sees "grading in progress.". Students can only see grades once published.

### publishAll, unpublishAll

A publishAll and unpublishAll feature has been added to the admin-only ground control to allow for bulk publishing / unpublishing.

image

^new field added to controls, 'Release Grading', with unpublish and publish all buttons added in the category (in that order)

image

^example prompt when 'publish all' button is clicked - checks are done in the backend to prevent publishing of non-graded work

### isAutoPublished

To avoid the inconvenience of publishing for trivial autograded assessments like Paths, a new backend field, 'isAutoPublished', has been added to each type of assessment, controllable within the admin panel. This field permits the autograder to publish on behalf of the avenger one it has fully graded the work. This field can only be set to true for assessments that are not manually graded (frontend and backend guards are present to prevent manually graded assessments from being auto-published, and vice-versa).

image

^new toggle, "is Auto-published", has been introduced.

### Revamp derived frontend field

With the introduction of publishing for submissions, multiple backend fields can now be revamped and combined into a single comprehensive 'progress' state - you would have observed this in the screenshots above as the single column, 'progress', for the grading submissions table.

This new field is calculated as such:

| isManuallyGraded |     status    | is fully graded? | is grading published? |    progress   |
|------------------|:-------------:|:----------------:|:-------------:|:-------------:|
| false            | ?             | ?                | ?             | autograded    |
| true             | attempting    | ?                | ?             | attempting    |
| true             | attempted     | ?                | ?             | attempted     |
| true             | submitted     | no               | ?             | submitted     |
| true             | submitted     | yes              | no            | graded        |
| true             | submitted     | yes              | yes           | published     |

Multiple backend parameters are considered and combined into a single field, allowing avengers to easily track and filter the submissions they need to grade under a single column.

image

^old control flow - credit to PR description in #2522

image

^old appearance of grading table - notice that 2 columns are used to track each assessment.

### Remove avenger need to track autograded work

newAvengerWorkflow

^new control flow for avengers.

All submissions which are part of autograded work has been marked as 'autograded'. Under our current changes, we expect 2 scenarios, both of which do not require the TA to be involved.

- Autograded, autopublished assessments (e.g. Paths):  They should not be unsubmitted or published by the TA. It is unfair since the marking scheme will be automatically revealed - besides, it should be too trivial to be cared about.
- Autograded, manually published work (e.g. practical exams): They are too serious and should not be able to be unsubmitted or published by the avenger. Only the admins should be able to publish autograding all at once using Ground Control.

Hence, our default filter will hide autograded work and disable actions for autograded work. Avengers will only be able to look through submissions and adjust XP if absolutely needed (see the lack of grading actions available for autograded work above).

###  Notification message change

image

^Old notifications show "autograded" notification even when not yet published.

image

^New notifications do not show any notification until published.

The notifications for autograded assessments that are not automatically published has become misleading because of our change - autograded submissions immediately trigger a notification even when not yet published. Thus, the frontend has been tweaked in accordance to the notification type changes in https://github.com/source-academy/backend/pull/965 so that only published and unpublished notifications are given.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

This pull request relies on logic that are derived from multiple additional non-optional fields introduced in https://github.com/source-academy/backend/pull/965 ('isGradingPublished' field for a submission and 'isAutoPublished' field for each assessment type), and would break if not provided.

- [X] This change requires a documentation update

Notification types and grading submission tables have been revamped

- [] Code quality improvements

### Checklist

- [X] I have tested this code
- [ ] I have updated the documentation

### Known issue

image

The dashboard that monitors grading progress of avengers has also become misleading, as it simply checks for graded assignments (that may not be published). It will be more accurate for this to instead (or additionally) indicate the number of published submissions so that the dashboard can properly reflect the feedback that can be received by students. To prevent this PR from getting any bigger, this has been shifted as a future issue to be addressed under the paired issues https://github.com/source-academy/frontend/issues/2898 and https://github.com/source-academy/backend/issues/1099.

### CODE REVIEW

**Comment by josh1248** [comment]
marked as draft to implement all features regarding isPublished within this pull

File: src/pages/academy/groundControl/subcomponents/GroundControlPublishCell.tsx
Comment (RichDom2185) [inline src/pages/academy/groundControl/subcomponents/GroundControlPublishCell.tsx:9]: Please fix this typo

**Review by RichDom2185 — COMMENTED** [review]
Hi, thanks for working on this! I've looked through ~half of the PR, preliminary comments below. Thanks!

File: src/pages/academy/dashboard/Dashboard.tsx
Comment (josh1248) [reply src/pages/academy/dashboard/Dashboard.tsx:64]: not yet

File: src/pages/academy/grading/subcomponents/GradingActions.tsx
Comment (josh1248) [reply src/pages/academy/grading/subcomponents/GradingActions.tsx:52]: done - removed the comment!

File: src/pages/academy/groundControl/subcomponents/GroundControlPublishCell.tsx
Comment (josh1248) [reply src/pages/academy/groundControl/subcomponents/GroundControlPublishCell.tsx]: done!

File: src/pages/academy/dashboard/Dashboard.tsx
Comment (josh1248) [reply src/pages/academy/dashboard/Dashboard.tsx:64]: yep, this change requires some small tweaks but this PR is getting too bloated - I think we will convert this to a good first issue.

Comment (martin-henz) [reply src/pages/academy/dashboard/Dashboard.tsx:64]: OK, once you have made the issue, please change the comment so that we can resolve this conversation.

Comment (josh1248) [reply src/pages/academy/dashboard/Dashboard.tsx:64]: paired issue on the frontend and backend has been opened. will resolve this now.

Comment (RichDom2185) [reply src/pages/academy/dashboard/Dashboard.tsx:64]: Perhaps you can link the issue in the comment?

Comment (josh1248) [reply src/pages/academy/dashboard/Dashboard.tsx]: alright, here are the issues:
Frontend issue: https://github.com/source-academy/frontend/issues/2898
Backend issue: https://github.com/source-academy/backend/issues/1099

this change needs both frontend and backend issues, so i have opened them up together as paired issues. the issues will be linked to this comment thread.

File: src/pages/academy/groundControl/GroundControl.tsx
Comment (josh1248) [inline src/pages/academy/groundControl/GroundControl.tsx:232]: changes to this and the row below that increases row height were meant to mitigate UI issues from this change since an additional column was added into ground control. However, these can be ignored if the ground control improvements in #2881 will be able to resolve such UI issues.

File: src/commons/sagas/RequestsSaga.ts
Comment (RichDom2185) [inline src/commons/sagas/RequestsSaga.ts:432]: Is this TODO still relevant?

File: src/commons/sagas/RequestsSaga.ts
Comment (RichDom2185) [inline src/commons/sagas/RequestsSaga.ts:486]: Ditto.

File: src/commons/sagas/RequestsSaga.ts
Comment (josh1248) [reply src/commons/sagas/RequestsSaga.ts]: yep, I forget to remove the frontend field calculation even though it is no longer needed for the profile display... Removed it now!

File: src/commons/sagas/RequestsSaga.ts
Comment (josh1248) [reply src/commons/sagas/RequestsSaga.ts]: Removed :)

File: src/commons/application/actions/SessionActions.ts
Comment (josh1248) [reply src/commons/application/actions/SessionActions.ts]: corrected! Thanks for catching that :D

**Review by RichDom2185 — COMMENTED** [review]
Hi, thanks for working on this!

LGTM, but will wait for backend PR to pass review before approving and merging both together.

**Review by RichDom2185 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
